relax_spec.md

```diff
@@ -0,0 +1,803 @@
+         | Var(name_hint: string, struct_info_annotation: StructInfo?)
+           # scoped to DataflowBlocks
+         | DataflowVar(name_hint: string, struct_info_annotation: StructInfo?)
+         | GlobalVar(name_hint: string)
+         | Tuple(fields: [Expr])
+         | SeqExpr(blocks: [BindingBlock], body: Expr)
+         | PrimValue(value: PrimExpr)
+         | StringImm(value: string)
+         | DataTypeImm(value: DataType)
+         | Function(params: [Var], body: Expr, ret_struct_info: StructInfo?, attrs: Attrs?)
+         | If(cond: Expr, true_branch: Expr, false_branch: Expr)
+         | ExternFunc(global_symbol: string)
+         | Call(op: Expr, args: [Expr], sinfo_args: [StructInfo], attrs: Attrs?)
+         | ShapeExpr(values: [PrimExpr])
+         | TupleGetItem(tuple_value: Expr, index: int)
+         | Op(op_name: string)
+
+# binding blocks (analogous to sequence of statements)
+BindingBlock ::= 
+           BindingBlock(bindings: [Binding])
+         | DataflowBlock(bindings: [Binding])
+
+# bindings (analogous to statements)
+Binding ::= 
+           VarBinding(var: Var|DataflowVar, value: Expr)
+         | MatchCast(var: (Var|DataflowVar)?, struct_info: StructInfo, value: Expr)
+
+# Relax programs are IRModules. Modules may bind global variables either to
+# Relax functions or TIR PrimFuncs (specified separately).
+# The Relax compiler may analyze and modify the TIR PrimFUncs as well.
+Program ::= IRModule(funcs: {GlobalVar: Function|PrimFunc})
+```
+
+### Notes on `derive_func`
+
+The `derive_func` field of `FuncStructInfo` is a macro in the meta-language: Given a function call and the variable mapping context, return the `StructInfo` of the result. This field is used only at compile time for reasoning about the `StructInfo` of calls to `ExternFunc`s.
+
+### Notes on `DataType` and Related Terminology
+
+The representation of datatypes, `DataType`, in the above AST is taken directly from TIR. However, the usage of datatypes in Relax is more restricted than in TIR.
+1. The `lanes` field for the `Int`, `UInt`, and `Float` datatypes must always be 1; we do not directly consider vectorized values in Relax.
+2. The `lanes` field for the `Handle` datatype must always be 0, indicating that it is `Void` (see below). The `bits` field for `Handle` should always be set to 64 (it will not be used by Relax).
+
+We also define the following special notation for datatypes, to be used in the rest of the specification:
+1. `Bool()`: This is shorthand for `UInt(bits=1, lanes=1)`, since TIR does not have a separate Boolean type. "True" refers to a value of 1 in this datatype and "false" refers to a value of 0. For convenience, we will refer to Boolean values as a separate datatype in the specification, due to their significance in `If` nodes.
+2. `Void()`: This is shorthand for `Handle(bits=64, lanes=0)`. TIR uses this datatype to refer to opaque objects; in Relax, it is used to denote an unknown datatype.
+
+## Expression Survey
```

Comment:
We need the AST spec here.  I find the description below confusing.  For example `PrimExpr` vs `PrimValue`...

Reply:
Would you recommend including the structure of the nodes in the below description? The AST diagram is in the section above, so I'm not sure where I would move it.

Reply:
No, it was a brainfart.  I was focusing on reading this section and I forgot what I read 10 minutes prior...  :|